[Kernel] ContentManager: add support for STFS packages

emoose opened the pull request:
This PR allows ContentManager to see and mount from non-extracted STFS packages. It'll also read information from the package headers (such as display name) into the XCONTENT_DATA struct used during content enumeration (see #1407)

I also added some code to allow extracted packages to have a [filepath].headers file next to it with the StfsHeaders struct inside, so they can also have information read into XCONTENT_DATA, but I could remove this part from the PR if desired.

gibbed commented on src/xenia/kernel/xam/content_manager.cc:
```diff
@@ -247,7 +290,11 @@ X_RESULT ContentManager::DeleteContent(const XCONTENT_DATA& data) {
   auto package_path = ResolvePackagePath(data);
   if (xe::filesystem::PathExists(package_path)) {
-    xe::filesystem::DeleteFolder(package_path);
+    if (xe::filesystem::IsFolder(package_path)) {
```
Okay. Conditional logic like this is a bad idea, perhaps we need to make a base class and derive into multiple classes for a folder/package/etc?

emoose replied:
Yeah that would be a lot better, I'll have a try seeing if I can get something like that to work now.

emoose commented on src/xenia/kernel/xam/content_manager.cc:
```diff
@@ -35,8 +36,19 @@ ContentPackage::ContentPackage(KernelState* kernel_state, std::string root_name,
+  // Otherwise mount as a local host path
+  if (!filesystem::IsFolder(package_path) &&
+      filesystem::PathExists(package_path)) {
+    device =
```
I was thinking the same, I'll try moving most of the code into a separate Mount method instead.

Nikota88 replied:
> Not sure I like this, failure conditions in a constructor are bad.
> 
> 
> 
> Maybe we should make the constructor private and expose a public 'constructor' method?

emoose commented on src/xenia/kernel/xam/content_manager.cc:
```diff
@@ -107,15 +124,41 @@ std::vector<XCONTENT_DATA> ContentManager::ListContent(uint32_t device_id,
+
+    auto headers_path = file_info.path + file_info.name;
+    if (file_info.type == xe::filesystem::FileInfo::Type::kDirectory) {
+      headers_path = headers_path + L".headers";
```
True, ```.headers.bin``` should be fine.

Also I think Velocity should be able to let people edit these header-only files, but don't quote me on that, if not maybe I'll take a look at that soon too...

emoose commented:
Had a try at making ContentPackage a base class which StfsContentPackage & FolderContentPackage inherit from, the code seems to work fine, with both package types loading in mixed-together without issue.

Also changed how the packages are managed a little, instead of ContentManager holding a std::map of root_path -> package pairs, it's now a vector of packages, with each containing their own root_path vectors instead.

This should let a single package be mounted under multiple root paths (in case that's ever needed), with the ContentPackage instance also handling the symlinking/desymlinking for each path.

emoose commented:
I've done a big rewrite of the STFS structures using things like xe::be, so that those structs will work just like the other X360 structs we use in Xenia, without needing any StfsHeader::Read/StfsHeader::Write functions, now we can just cast a pointer to it, or copy the memory of it directly, etc.

Also included a few header structures we were missing like the Installer struct, which is used on title update packages to specify some info about the title being patched, along with a couple other ones.

Wasn't sure about adding it to this PR though since it seems this might need some changes already, so I've put that rewrite into another branch for now:
https://github.com/emoose/xenia/commits/stfs-headers

Still need to test that and make sure I haven't broken anything, hopefully it _should_ be working fine though.

TranzRail commented:
Unfortunately I didn't have a SVOD to try with, but I did merge your changes with the master on a local repo and didn't notice any ill effects. In fact, LEGO Rock Band was able to properly read (and write) the game data (via XamUserReadProfileSettings), whereas I think I was having issues with it before. The common storage issue I see in the logs for the games I test with is RtlNtStatusToDosError 103 => 3E5, which I've noted before.

emoose commented:
Added some code in [stfs-headers](https://github.com/emoose/xenia/commits/stfs-headers) to make ContentManager::CreateContent write the saves display name/content type out to a headers.bin file, this should now allow games like Kameo & Halo 3 to recognise their own save games!
(also rebased that branch to latest master too)

@TranzRail good to hear, interesting that it helped with XamUserReadProfileSettings, don't think this should touch any of that, guess that maybe it needed to do something save/STFS related for the profile settings stuff to work properly.

The RtlNtStatusToDosError messages seem pretty common, looks like a lot of games try to access files that don't exist, think in most cases it isn't really a fault with Xenia, probably more a fault with lazy devs :P

TranzRail commented:
@emoose The error in question is ERROR_IO_PENDING (Overlapped I/O operation is in progress) when trying to write/save. So I doubt that'll be because of lazy devs, given how common it is? 

The XamUserReadProfileSettings I think could do with a revamp at some point. Have it more structured.

It's possible that in some cases, some games will be looking for a particular mount that does not exist. I don't know how it currently works, but I imagine we'd want to setup our own static mounts for things that would exist when dash.xex is booted for instance (for things like internal storage, optical drive, flash, USB, etc) and then we can middle manage them between the exported functions.

illusion98 commented:
@emoose just tested Dead Space 1 with https://github.com/emoose/xenia/commit/7e2a163a51836be07d54959aee134a32b7ffd169 in SVOD and it loaded fine

emoose commented:
Saw this mentioned on twitter, I started another STFS branch at https://github.com/emoose/xenia/commits/stfs-improvements a little while ago, mostly just improvements for STFS structs & some fixes for CON packages (atm it looks like current code doesn't work with them), a lot less involved than this PR, will try PRing it soon.

I'm not sure if the method used in this PR with different StfsContentPackage/FolderContentPackage classes is really that ideal anymore, maybe it'd be better to have the ContentPackage class use StfsContainerDevice, let that device handle the STFS header stuff for us, but also let it create/own a HostPathDevice itself for folder packages if needed, that way all folder packages will have STFS headers handled for them (for any XamContent stuff that need to access them), and would also let both folders + packages use mostly the same code.

I think StfsContainerDevice owning a HostPathDevice might be the best way to allow loading in CON-packaged saves/profiles while also letting changes be written back to disk too, unless someone wants to try adding write support to the STFS code... not really a fan of writing things to binary blobs that need tools to access though.
